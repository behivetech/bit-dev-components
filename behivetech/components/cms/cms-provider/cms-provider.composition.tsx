import * as React from 'react';
import { CmsProvider, useCmsProvider } from './cms-provider';

const Header = ({headerContent = 'Default Header', ...rest}) => (
    <header>
        <h1>{headerContent}</h1>
    </header>
);

const Footer = ({footerContent = 'Default Footer'}) => (
    <footer>
        {footerContent}
    </footer>
);

const Main = ({children = null, mainContent = "Default Main"}) => (
    <main>
        {children}
        {mainContent}
    </main>
);

const AddedBlock = ({children = null}) => (
    <div data-testid="addedBlock">{children}</div>
);

const MockCmsArea = ({name}) => {
    const {cmsParams: {areas}} = useCmsProvider();
    const blocks = areas[name] ? areas[name].blocks : [];
    const blockComponents = {
        AddedBlock,
        Header,
        Footer,
        Main,
    }

    return (
        <>{
            blocks.length
                ? blocks.map(({componentName, componentProps}, index) => {
                    const Component = blockComponents[componentName];

                    return <Component key={`${componentName}__${index}`} {...componentProps} />;
                })
                : null
        }</>
    );
}

const ButtonActions = () => {
    const {
        addBlock,
        layoutDrawerDetails,
        cmsParams,
        removeBlock,
        setLayoutDrawerDetails,
        toggleEditBlock,
        updateBlockComponent,
    } = useCmsProvider();

    function handleAddBlock() {
        addBlock('main', 'AddedBlock', {children: 'Added Block'});
    }

    function handleEditBlock() {
        setLayoutDrawerDetails(<div>Edit Drawer Details</div>);
        toggleEditBlock();
    }

    function handleRemoveBlock() {
        removeBlock('main', 1);
    }

    function handleUpdateBlock() {
        updateBlockComponent('main', 0, {mainContent: 'Main Block Updated'});
    }

    return (
        <div data-testid="button-actions">
            <button data-testid="addBlock" onClick={handleAddBlock}>Add Block</button>
            <button data-testid="editBlock" onClick={handleEditBlock}>Toggle Edit Block</button>
            <button data-testid="removeBlock" onClick={handleRemoveBlock}>Remove Added Block</button>
            <button data-testid="updateBlock" onClick={handleUpdateBlock}>Update Block</button>
            {layoutDrawerDetails}
        </div>
    )
}

export const BasicCmsProvider = () => {
    const cmsParams = {
        areas: {
            header: {
                blocks: [{
                    componentName: 'Header',
                    componentProps: {headerContent: 'Header Component Content'}
                }],
            },
            footer: {
                blocks: [{
                    componentName: 'Footer',
                    componentProps: {footerContent: 'Footer Component Content'}
                }],
            },
            main: {
                blocks: [{
                    componentName: 'Main',
                    componentProps: {
                        children: (
                            <div>
                                <div>
                                    This is a composition of functionality and content to test out
                                    the CmsProvider component.
                                </div>
                                <ButtonActions />
                            </div>
                        ),
                        mainContent: 'Main Component Content'}
                    }],
            },
        }
    }

    return (
        <CmsProvider cmsParams={cmsParams}>
            <MockCmsArea name="header" />
            <MockCmsArea name="main" />
            <MockCmsArea name="footer" />
        </CmsProvider>
    );
};
