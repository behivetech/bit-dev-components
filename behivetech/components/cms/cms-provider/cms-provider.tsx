import * as React from 'react';

export type ComponentPropsType = {[propName: string]: any}

type BlockType = {
    /**
     * Name of component to be imported
     */
    componentName: string,
    /**
     * Props to be passed into the block component
     */
    componentProps: ComponentPropsType
}

export type BlocksType = {
    /**
     * an array which contains which component to use by name and data prop passed into the block component.
     */
    blocks: BlockType[]
}

type CmsParamsType = {
    /**
     * keys are the area name which contains the blocks that go within an area.
     */
    areas?: {
        /**
         * The name of an area that contains the blocks
         */
        [areaName: string]: BlocksType
    },
}

type ContextType = {
    /**
     * Adds a new block to an area
     */
    addBlock: (areaName: string, componentName: string, componentProps: ComponentPropsType) => void;
    /**
     * Determines if the CMS interface is in edit block mode or not which opens
     * a drawer with contents to update the block.
     */
    editBlock: boolean;
    /**
     * Determines if the CMS interface is in edit mode for the page or not.
     */
    editPage: boolean;
    /**
     * A react node of children to go within the layout drawer to edit a block
     */
    layoutDrawerDetails: React.ReactNode;
    /**
     * Params used by the CMS components to build the page
     */
    cmsParams: CmsParamsType
    /**
     * Removes a block from an area
     */
    removeBlock: (areaName: string, index: number) => void;
    /**
     * Sets the contents of the layout drawer to modify a block
     */
    setLayoutDrawerDetails: (details: React.ReactNode) => void;
    /**
     * Sets the editBlock to true/false
     */
    toggleEditBlock: (openOverride?: boolean) => void;
    /**
     * Sets the editPage to true/false
     */
    toggleEditPage: () => void;
    /**
     * Updates the dynamic content of a block
     */
    updateBlockComponent: (areaName: string, index: number, componentProps: ComponentPropsType) => void;
}

interface CmsCallbackInterface extends CmsParamsType {
    onFailure: () => void;
    onSuccess: (cmsParams?: CmsParamsType) => CmsParamsType | undefined;
};

export type CmsProviderProps = {
    /**
    * children to be rendered in the component.
    */
    children: React.ReactNode,
    /**
    * data used to define layouts, areas and blocks and determines where they go within the view.
    */
    cmsParams: CmsParamsType,
};

const DEFAULT_CMS_PARAMS = {areas: {}};
const DEFAULT_CONTEXT: ContextType = {
    addBlock: (areaName, componentName, componentProps) => null,
    editBlock: false,
    layoutDrawerDetails: null,
    editPage: false,
    removeBlock: (areaName, index) => null,
    setLayoutDrawerDetails: (details) => null,
    toggleEditPage: () => null,
    toggleEditBlock: (openOverride) => null,
    updateBlockComponent: (areaName, index, componentProps) => null,
    cmsParams: DEFAULT_CMS_PARAMS,
};

const CmsProviderContext = React.createContext(DEFAULT_CONTEXT);

export function useCmsProvider() {
    return React.useContext(CmsProviderContext)
};

function reducer(state = {areas: {}}, {type, values}) {
    function getBlocks(areaName: string): {[key: string]: any}[] {
        const stateArea = state.areas[areaName];

        return stateArea && stateArea.blocks ? stateArea.blocks : [];
    }

    function updateAreas(areaName, blocks) {
        return state.areas[areaName]
            ? {...state.areas, [areaName]: {blocks}}
            : state.areas
    }

    const actions = {
        ADD_BLOCK: () => {
            const {areaName, componentName, componentProps} = values;

            return {
                ...state,
                areas: updateAreas(areaName, [...getBlocks(areaName), {componentName, componentProps}]),
            };
        },
        REMOVE_BLOCK: () => {
            const {areaName, index} = values;
            const newBlocks = getBlocks(areaName);

            if (areaName && newBlocks && newBlocks.length) {
                newBlocks.splice(index, 1);
            }

            return {...state, areas: updateAreas(areaName, newBlocks)};
        },
        UPDATE_BLOCK_DATA: () => {
            const {areaName, index, componentProps} = values;
            const newBlocks = getBlocks(areaName);

            if (areaName && newBlocks && newBlocks.length) {
                const newBlock = newBlocks[index];

                newBlocks.splice(
                    index,
                    1,
                    {...newBlock, componentProps: {...newBlock.componentProps, ...componentProps}}
                );
            }

            return newBlocks.length ? {...state, areas: updateAreas(areaName, newBlocks)} : state;
        },
    };

    const newState = actions[type] ? actions[type]() : state;

    return newState;
}

export function CmsProvider({ children, cmsParams = DEFAULT_CONTEXT.cmsParams }: CmsProviderProps) {
    const [editBlock, setEditBlock] = React.useState(false);
    const [layoutDrawerDetails, setLayoutDrawerDetailsState] = React.useState(null);
    const [editPage, setEditPage] = React.useState(false);
    const [cmsState, dispatchPageState] = React.useReducer(reducer, cmsParams);

    function setLayoutDrawerDetails(details = null) {
        setLayoutDrawerDetailsState(details);
    }

    function removeBlock(areaName, index) {
        dispatchPageState({type: 'REMOVE_BLOCK', values: {areaName, index}});
        setEditBlock(false);
    }

    function addBlock(areaName, componentName, componentProps) {
        dispatchPageState({type: 'ADD_BLOCK', values: {areaName, componentName, componentProps}});
        setEditBlock(false);
    }

    function toggleEditPage() {
        setEditPage(!editPage);
        setEditBlock(false);
        setLayoutDrawerDetailsState(null);
    }

    function toggleEditBlock(openOverride) {
        const newDrawerOpen = openOverride !== undefined ? openOverride : !editBlock;

        setEditBlock(newDrawerOpen);

        if (!newDrawerOpen) {
            console.info('Clearing Drawer Layout');
            setLayoutDrawerDetailsState(null);
        }
    }

    function updateBlockComponent(areaName, index, componentProps) {
        dispatchPageState({type: 'UPDATE_BLOCK_DATA', values: {areaName, index, componentProps}});
        setEditBlock(false);
    }

    const context = {
        addBlock,
        editBlock,
        layoutDrawerDetails,
        editPage,
        cmsParams: cmsState,
        removeBlock,
        setLayoutDrawerDetails,
        toggleEditBlock,
        toggleEditPage,
        updateBlockComponent,
    };

    return (
        <CmsProviderContext.Provider value={context}>
            {children}
        </CmsProviderContext.Provider>
    );
}
