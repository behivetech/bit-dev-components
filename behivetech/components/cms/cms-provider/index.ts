export { CmsProvider, useCmsProvider } from './cms-provider';
export type { CmsProviderProps } from './cms-provider';
