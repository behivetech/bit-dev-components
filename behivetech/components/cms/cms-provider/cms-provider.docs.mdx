---
description: 'A CmsProvider component.'
labels: ['cms provider', 'react', 'provider']
---

import { CmsProvider } from './cms-provider';

## React Component for rendering a React Provider for React's context
This provider should wrap the entire app so all child CMS components can access the `cmsParams` and other functionality from the context passed down from this [React provider](https://reactjs.org/docs/context.html). For instance, in a Next.js project, it should utilize the [`_app.tsx`](https://nextjs.org/docs/advanced-features/custom-app) component which is utilized on every page in the Next.js framework. This component requires props for `children` and `cmsParams`. The `cmsParams` prop is an object that defines params to be used for layouts, areas, blocks and any component props needed for a block level component.

Here's a simple code block on how to use the components.

```jsx
// Custom components created by the developer
const Header = () => <h1 className="header">I'm a header</h1>
const Footer = () => <footer className="footer">I'm a footer</footer>
const MainTitle = () => <h2 className="main-title">I'm the Main Title</h2>
const Aside = ({sideNav}) => (
    <aside>
        <ul>
            {sideNav.map(({label, url}) => <li key={label}><a href={url}>{label}</a></li>)}
        </ul>
    </aside>
)
const Section = ({sectionNumber}) => <section>I'm a section {sectionNumber}</section>

// Objects defining the specific components needed within an area
const headerImportedComopnents = {Header}
const footerImportedComopnents = {Footer}
const mainImportedComopnents = {Section, Aside}

const cmsParams = {
    areas: {
        header: {blocks: [{componentName: 'Header'}]}
        footer: {blocks: [{componentName: 'Footer'}]}
        main: {blocks: [
            {componentName: 'Section', componentProps: {sectionNumber: 1}}
            {componentName: 'Section', componentProps: {sectionNumber: 2}}
            {componentName: 'Aside', componentProps: {sideNav: [
                {label: 'Link 1', url: 'https://www.somewhere1.com'}
                {label: 'Link 2', url: 'https://www.somewhere2.com'}
            ]}}
        ]}
    }
}

export default function MyPageComponent() {
    return (
        <CmsProvider cmsParams={cmsParams}>
            <div className="my-page-component">
                <header className="my-page-component__header">
                    <CmsArea name="header" importedComponents={headerImportedComopnents} />
                </header>
                <main className="my-page-component__main">
                    <CmsArea name="main" importedComponents={mainImportedComopnents} />
                </main>
                <CmsArea name="footer" importedComponents={footerImportedComopnents} />
            </div>
        </CmsProvider>
    );
}

```

In this example, everything is wrapped in the `CmsProvider` so the correct context can be passed down to the `CmsBlocks` component. The CmsArea component passes its name prop down to the `CmsBlocks` component which in turn will use that as a key to pull the params out of the CmsParams specific to that area. It will look at the blocks key which should contain an array of params for the Blocks. The component name tells the Block which component should be pulled from the object of imported components (headerImportedComopnents, footerImportedComopnents, mainImportedComopnents). The componentProps key passes in any props defined as its value. The above code should render the following HTML...

```html
<div class="my-page-component">
    <header>
        <h1 class="my-page-component__header">I'm a header</h1>
    </header>
    <main class="my-page-component__main">
        <section>I'm a section 1</section>
        <section>I'm a section 2</section>
        <aside>
            <ul>
                <li><a href="https://www.somewhere1.com">Link 1</a></li>
                <li><a href="https://www.somewhere2.com">Link 2</a></li>
            </ul>
        </aside>
    </main>
    <footer class="footer">I'm a footer</footer>
</div>
```

Typically the `cmsParams` would come from some sort of API (see upcomoing example project) to handle the CRUD operations for the CMS. In a Next.js project, the Static Generation function, `getStaticProps`,  or SSR function, `getServerSideProps` could be used at the page level in efforts to reduce network calls and less work for the client side browser.
