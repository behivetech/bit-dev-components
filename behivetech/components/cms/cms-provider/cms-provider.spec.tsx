import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { BasicCmsProvider } from './cms-provider.composition';

describe('CmsProvider',  () => {
    it('should render with the correct content', () => {
        const {asFragment} = render(<BasicCmsProvider />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should add a block', () => {
        const {asFragment} = render(<BasicCmsProvider />);

        fireEvent.click(screen.getByTestId('addBlock'));
        expect(asFragment()).toMatchSnapshot();
    })

    it('should remove a block', () => {
        const {asFragment} = render(<BasicCmsProvider />);

        fireEvent.click(screen.getByTestId('removeBlock'));
        expect(asFragment()).toMatchSnapshot();
    })

    it('should update a block', () => {
        const {asFragment} = render(<BasicCmsProvider />);

        fireEvent.click(screen.getByTestId('updateBlock'));
        expect(asFragment()).toMatchSnapshot();
    })

    it('should render edit content', () => {
        const {asFragment} = render(<BasicCmsProvider />);

        fireEvent.click(screen.getByTestId('editBlock'));
        expect(asFragment()).toMatchSnapshot();
    })
})
