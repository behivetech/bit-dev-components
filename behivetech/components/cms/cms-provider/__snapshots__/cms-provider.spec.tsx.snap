// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CmsProvider should add a block 1`] = `
<DocumentFragment>
  <header>
    <h1>
      Header Component Content
    </h1>
  </header>
  <main>
    <div>
      <div>
        This is a composition of functionality and content to test out the CmsProvider component.
      </div>
      <div
        data-testid="button-actions"
      >
        <button
          data-testid="addBlock"
        >
          Add Block
        </button>
        <button
          data-testid="editBlock"
        >
          Toggle Edit Block
        </button>
        <button
          data-testid="removeBlock"
        >
          Remove Added Block
        </button>
        <button
          data-testid="updateBlock"
        >
          Update Block
        </button>
      </div>
    </div>
    Main Component Content
  </main>
  <div
    data-testid="addedBlock"
  >
    Added Block
  </div>
  <footer>
    Footer Component Content
  </footer>
</DocumentFragment>
`;

exports[`CmsProvider should remove a block 1`] = `
<DocumentFragment>
  <header>
    <h1>
      Header Component Content
    </h1>
  </header>
  <main>
    <div>
      <div>
        This is a composition of functionality and content to test out the CmsProvider component.
      </div>
      <div
        data-testid="button-actions"
      >
        <button
          data-testid="addBlock"
        >
          Add Block
        </button>
        <button
          data-testid="editBlock"
        >
          Toggle Edit Block
        </button>
        <button
          data-testid="removeBlock"
        >
          Remove Added Block
        </button>
        <button
          data-testid="updateBlock"
        >
          Update Block
        </button>
      </div>
    </div>
    Main Component Content
  </main>
  <footer>
    Footer Component Content
  </footer>
</DocumentFragment>
`;

exports[`CmsProvider should render edit content 1`] = `
<DocumentFragment>
  <header>
    <h1>
      Header Component Content
    </h1>
  </header>
  <main>
    <div>
      <div>
        This is a composition of functionality and content to test out the CmsProvider component.
      </div>
      <div
        data-testid="button-actions"
      >
        <button
          data-testid="addBlock"
        >
          Add Block
        </button>
        <button
          data-testid="editBlock"
        >
          Toggle Edit Block
        </button>
        <button
          data-testid="removeBlock"
        >
          Remove Added Block
        </button>
        <button
          data-testid="updateBlock"
        >
          Update Block
        </button>
        <div>
          Edit Drawer Details
        </div>
      </div>
    </div>
    Main Component Content
  </main>
  <footer>
    Footer Component Content
  </footer>
</DocumentFragment>
`;

exports[`CmsProvider should render with the correct content 1`] = `
<DocumentFragment>
  <header>
    <h1>
      Header Component Content
    </h1>
  </header>
  <main>
    <div>
      <div>
        This is a composition of functionality and content to test out the CmsProvider component.
      </div>
      <div
        data-testid="button-actions"
      >
        <button
          data-testid="addBlock"
        >
          Add Block
        </button>
        <button
          data-testid="editBlock"
        >
          Toggle Edit Block
        </button>
        <button
          data-testid="removeBlock"
        >
          Remove Added Block
        </button>
        <button
          data-testid="updateBlock"
        >
          Update Block
        </button>
      </div>
    </div>
    Main Component Content
  </main>
  <footer>
    Footer Component Content
  </footer>
</DocumentFragment>
`;

exports[`CmsProvider should update a block 1`] = `
<DocumentFragment>
  <header>
    <h1>
      Header Component Content
    </h1>
  </header>
  <main>
    <div>
      <div>
        This is a composition of functionality and content to test out the CmsProvider component.
      </div>
      <div
        data-testid="button-actions"
      >
        <button
          data-testid="addBlock"
        >
          Add Block
        </button>
        <button
          data-testid="editBlock"
        >
          Toggle Edit Block
        </button>
        <button
          data-testid="removeBlock"
        >
          Remove Added Block
        </button>
        <button
          data-testid="updateBlock"
        >
          Update Block
        </button>
      </div>
    </div>
    Main Block Updated
  </main>
  <footer>
    Footer Component Content
  </footer>
</DocumentFragment>
`;
